_posts/2017-03-30-storage-formats-quickstep.markdown

```diff
@@ -0,0 +1,140 @@
+---
+layout: post
+title:  "Storage Formats in Quickstep"
+date: 2017-03-30 14:00:30 -0600
+categories: guides
+author: Harshad
+short: Quickstep offers variety of storage formats. This post explains what these formats are and how to use them.
+---
+One of the strengths of Quickstep is the variety of storage formats it offers to store the relational data. The storage management in Quickstep can work with all these formats, and each format comes with its own strengths and weaknesses. The foundation of this work was laid in a [research paper](http://www.vldb.org/pvldb/vol6/p1474-chasseur.pdf) from the Patel Wisconsin Database group that appeared in [VLDB 2014](http://vldb.org/) (International Conference on Very Large Databases). In this post, I will provide a brief primer on what these storage formats mean and how to use them in the Quickstep system. I will also provide links to the relevant source code files, so that the readers can start exploring the code base.
+
+<h1>Storage Formats 101</h1>
+A storage format refers to how the data in a table is laid out in the memory. Let's consider a toy relational schema.
+
+{% highlight sql %}
+CREATE TABLE employee (ID INT, Age INT, Name VARCHAR(8));
+{% endhighlight %}
+
+A tuple in the employee table consists of three attributes, id (an integer), age (an integer) and the name (character of 8 bytes fixed length). Let's also populate our table with few tuples. For simplicity, let us assume that the name is a fixed length attribute; however, in general it could be variable length. 
+
+{% highlight sql %}
+INSERT into employee values (101, 30, Jennifer);
+INSERT into employee values (102, 25, Jim);
+INSERT into employee values (103, 35, David);
+{% endhighlight %}
+
+Conceptually, our table looks as follows:
+
+| ID 	| Age 	|   Name   	|
+|:--:	|:---:	|:--------:	|
+|  101 	|  30 	| Jennifer 	|
+|  102 	|  25 	|    Jim   	|
+|  103 	|  35 	|   David  	|
+
+Next, let us dig into understanding the different ways in which this table can be stored in memory. 
+
+<h2>Row Store</h2>
+
+The first storage format we will look at is *row store*, which is a popular format, and one that is intuitive too.
+
+The figure below represents how a row store format for our toy example looks like in memory. The offsets indicate how far is the memory location from the beginning of the block. 
+
+![Row Store](../assets/storage-format-row-store.jpg)
+
+In the row store format, a tuple is stored in the table definition order. In memory, the ID (101) of the first tuple is stored at some address, and it is followed by the age field for that tuple (30), and then the name field ("Jennifer"). The second tuple is stored in memory sequentially after the first tuple, and so on.
```

Comment:
I agree only if it were fixed-length which I think it will not be. I think CHAR type may be fixed length but I have to go check in the code to see if it is implemented that way. if it is variable length, then it will be stored in a heap structure which grows backward from the end of the storage block. There will be a fixed length minituple in the fixed length zone for each variable length attribute. This fixed length minituple has the offset and length of the variable length string which the storage manager uses to retrieve the string from the heap